### PR TITLE
Group routes

### PR DESCRIPTION
**Pull request by kalebdavis**
fixes Pagernation/Pagernation#57

### CODE REVIEW

File: server/tests/routes/group.test.js [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+        users: [],
+        escalationPolicy: {}
+      };
+
+      it('should not create a new group', (done) => {
+        request(app)
+          .post(groupUrl)
+          .send(group)
+          .expect(httpStatus.BAD_REQUEST)
+          .then((res) => {
```
Comment (jarrydlee) [inline server/tests/routes/group.test.js]: You can one line this

File: server/models/group.js [hunk 2]
```diff
@@ -20,4 +27,24 @@ const GroupSchema = new mongoose.Schema({
...
+   * @param {GroupName} - the name of the group
+   * @return {Promise<Group, APIError>}
+   */
+
+  get(name) {
+    return this.findOne({ name: name })
+      .exec()
+      .then((group) => {
+        if (group) {
+          return group;
```
Comment (jarrydlee) [inline server/models/group.js]: I see that the `.exec()` will return a promise, but how does this line work once you return it? Will later `.then(...)` blocks be able to be chained together?

ie. 
``` javascript
Group.get('name')
    .then(name => console.log(name));
```

File: server/routes/validation/group.js [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+  },
+
+  byName: {
+    params: {
+      groupName: Joi.string().required()
+    }
+  },
+
+  updateGroup: {
+    options: { allowUnknownBody: false },
```
Comment (jarrydlee) [inline server/routes/validation/group.js]: Do you want the other validations to fail with invalid keys passed in with the object?

**Comment by kalebdavis** [comment]
The eslintrc is being changed because it really is just a JSON file, and this way it has syntax highlighting

File: server/routes/group.js [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+  // gets the group with the specified group name
+  .get(validate(paramValidation.byName), groupCtrl.getGroup)
+
+  // updates the group with the specified group name
+  .put(validate(paramValidation.updateGroup), groupCtrl.updateGroup)
+
+  .delete(validate(paramValidation.byName), groupCtrl.deleteGroup);
+
+// Loads user when API with groupName parameter is hit
```
Comment (jarrydlee) [inline server/routes/group.js]: Docs are wrong here

File: server/tests/models/group.test.js [hunk 5]
```diff
@@ -7,6 +7,11 @@ describe('Groups', () => {
   const subscriberObjectId = '57e590a0140ebf1cc48bb1bf';
 
   const userObjectId = '67e590a0140ebf1cc48bb1bf';
+  let group = null;
+
+  beforeEach(() => {
+    Group.remove({}, () => {});
```
Comment (jarrydlee) [inline server/tests/models/group.test.js]: Is this needed with the test changes that @paradime made? The DB should be dropped after each test now

File: server/tests/routes/group.test.js [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+          name: 1,
+          users: []
+        };
+
+        it('should not create a new group', (done) => {
+          request(app)
+            .post(groupUrl)
+            .send(group)
+            .expect(httpStatus.BAD_REQUEST)
+            .then((res) => {
```
Comment (jarrydlee) [inline server/tests/routes/group.test.js]: You should one line this

File: server/tests/routes/group.test.js [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+  context('with a group needed beforehand', () => {
+    context('with a valid group', () => {
+      const group = {
+        name: 'Wondertwins',
+        users: ['5c15f987046b1686d22dbea1', 'ba421976ad9b71458d8b91ab']
+      };
+
+      beforeEach((done) => {
+        groupService.createGroup(group)
+          .then((createdGroup) => {
```
Comment (jarrydlee) [inline server/tests/routes/group.test.js]: Same here

File: server/tests/services/group.test.js [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+
+describe('# Group Service', () => {
+  const groupObject = {
+    name: 'Wondertwins',
+    users: ['123456789012345678901234']
+  };
+
+  context('with no groups in the database beforehand', () => {
+    beforeEach(() => {
+      Group.remove({}, () => {});
```
Comment (jarrydlee) [inline server/tests/services/group.test.js]: Same comment about Byron's changes

File: server/tests/services/group.test.js [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+              expect(createdGroup.name).to.equal(groupObject.name);
+              expect(createdGroup.users[0].toString()).to.equal(groupObject.users[0]);
+              done();
+            })
+            .catch(err => done(err));
+        });
+      });
+
+      context('with missing user details', () => {
+        const missingName = {
```
Comment (jarrydlee) [inline server/tests/services/group.test.js]: nit: can you add specifics here? `missingNameGroup` just so you can know what it is without looking around

File: server/routes/group.js (hunk 4 above)
Comment (jarrydlee) [inline server/routes/group.js]: Loads _user_ when API ...

File: server/tests/routes/group.test.js [hunk 10]
```diff
@@ -0,0 +1,159 @@
+import request from 'supertest-as-promised';
+import httpStatus from 'http-status';
+import app from '../../../index';
+import groupService from '../../services/group';
+import Group from '../../models/group';
+
+const groupUrl = '/api/v1/group';
+
+describe('## Group API', () => {
```
Comment (paradime) [inline server/tests/routes/group.test.js]: this has been bothering me, we should standardize the `#` here

**Comment by paradime** [comment]
👍

**Pull request merged**